### PR TITLE
media-gfx/gimp: 2.99.12 version bump (dev preview)

### PR DESCRIPTION
**Pull request by band-a-prend**
In this release of Gimp dev-preview 2.99.12 the porting
from Autotools to Meson build system take place.

The Meson now is preferred by upstream.

Due to differences of configuration options withon configure.ac
and meson_options.txt the some src_configuration options dropped

1. related to flags:
`USE="aqua cpu_flags_ppc_altivec cpu_flags_ppc_mmx cpu_flags_ppc_sse"`

2. autotools configure options:
```
GEGL="${EPREFIX}"/usr/bin/gegl-0.4
GDBUS_CODEGEN="${EPREFIX}"/usr/bin/gdbus-codegen
--with-pdbgen
```

Also some patches are dropped in src_prepare() phase:
```
sed -i -e 's/== "xquartz"/= "xquartz"/' configure.ac || die #494864
sed 's:-DGIMP_DISABLE_DEPRECATED:-DGIMP_protect_DISABLE_DEPRECATED:g' -i configure.ac || die #615144

sed 's:-DGIMP_protect_DISABLE_DEPRECATED:-DGIMP_DISABLE_DEPRECATED:g' -i configure || die #615144
grep -F -q GIMP_DISABLE_DEPRECATED configure || die #615144, self-test
```
The `sed -i -e 's/== "xquartz"/= "xquartz"/' configure.ac || die #494864` actually doesn't do anything several releases.  

The `USE="X"` adds conditional support of `x11-libs/libXcursor`.

The FEATURES="test" now use `x11-misc/xvfb-run` instead virtualx.eclass to run tests.

In addition USE="doc" now build related to `gobject-introspection` (g-ir-doc)
plugin documentations. Early it's built was broken for autotools.  

--------------------

The main configuration issue is that [gimp meson.build (lines 178-198)](https://gitlab.gnome.org/GNOME/gimp/-/blob/master/meson.build) states
```
# Compiler CPU extensions for optimizations
#
# -- Don't enable these flags project-wide, only files needing them
# should be built with a given extension, which we do with meson simd
# module. Otherwise GIMP would end up crashing when run on CPUs which
# don't support all the flags supported by the compiler.
# See merge request !262 for more details.

# Note that Meson has a SIMD module which can also do this for us, but it uses
# its own #defines and we want to stay compatible with the autotools build
conf.set('USE_MMX', cc.has_argument('-mmx'))
conf.set('USE_SSE', cc.has_argument('-sse'))
conf.set10('COMPILE_SSE2_INTRINISICS', cc.has_argument('-msse2'))
conf.set10('COMPILE_SSE4_1_INTRINISICS', cc.has_argument('-msse4.1'))

if host_cpu_family == 'ppc'
  altivec_args = cc.get_supported_arguments([
    '-faltivec',
    '-maltivec',
    '-mabi=altivec',
  ])

```
And it's seems that MMX and SSE always switched off.  I got disabled them during meson configuration while they were enabled with autotools. The SSE2 and SSE4_1 were switched on during configuration.

**Should I left it as it is?**

The Altivec(ppc) activates automatically.

--------------------

On media-libs/gegl: 0.4.38 version bump (lto and sdl2):

Closes: https://bugs.gentoo.org/859901
Closes: https://bugs.gentoo.org/860246

P.S.
To add Python 3.11 I need to additionally test `dev-libs/gobject-introspection` after add Python 3.11 for it and changes  `PYTHON_REQ_USE="xml` to `PYTHON_REQ_USE="xml(+)"` and then `media-libs/gegl` and `media-gfx/gimp` again.
If it's OK then I will add additional commit.

### CODE REVIEW

File: media-gfx/gimp/gimp-2.99.12.ebuild [hunk 1]
```diff
@@ -0,0 +1,256 @@
+# Copyright 1999-2022 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+LUA_COMPAT=( luajit )
+PYTHON_COMPAT=( python3_{8..10} )
+GNOME2_EAUTORECONF=yes
```
Comment (thesamesam) [inline media-gfx/gimp/gimp-2.99.12.ebuild:8]: Drop.

File: media-gfx/gimp/gimp-2.99.12.ebuild [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+VALA_MIN_API_VERSION="0.50"
+VALA_USE_DEPEND=vapigen
+
+inherit lua-single meson python-single-r1 toolchain-funcs vala xdg
+
+DESCRIPTION="GNU Image Manipulation Program"
+HOMEPAGE="https://www.gimp.org/"
+SRC_URI="mirror://gimp/v2.99/${P}.tar.xz"
```
Comment (thesamesam) [inline media-gfx/gimp/gimp-2.99.12.ebuild:16]: New line after SRC_URI

File: media-gfx/gimp/gimp-2.99.12.ebuild [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+RDEPEND="
+	${COMMON_DEPEND}
+	x11-themes/hicolor-icon-theme
+	gnome? ( gnome-base/gvfs )
+"
+
+DEPEND="
+	${COMMON_DEPEND}
+	>=dev-lang/perl-5.30.3
```
Comment (thesamesam) [inline media-gfx/gimp/gimp-2.99.12.ebuild:95]: I think all of these are actually BDEPEND (after COMMON_DEPEND).

Also, you don't need autoconf/automake/libtool anymore. Maybe not even gettext & libtool (check).

File: media-gfx/gimp/gimp-2.99.12.ebuild [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+	# Fix Gimp  and GimpUI devel doc installation paths
+	sed -i -e "s/'doc'/'gtk-doc'/" devel-docs/reference/gimp/meson.build || die
+	sed -i -e "s/'doc'/'gtk-doc'/" devel-docs/reference/gimp-ui/meson.build || die
+
+	# Fix pygimp.interp python implementation path.
+	# Meson @PYTHON_PATH@ use sandbox path e.g.:
+	# '/var/tmp/portage/media-gfx/gimp-2.99.12/temp/python3.10/bin/python3'
+	sed -i -e 's:@PYTHON_PATH@:'${EPYTHON}':' plug-ins/python/pygimp.interp.in || die
+
+	export CC_FOR_BUILD="$(tc-getBUILD_CC)"
```
Comment (thesamesam) [inline media-gfx/gimp/gimp-2.99.12.ebuild:139]: You may not need this anymore.

File: media-gfx/gimp/gimp-2.99.12.ebuild [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+			# NOTE: Folder and file name need to match for Gimp to load that plug-in
+			#       so "file-svg/file-svg" becomes "${prename}file-svg/${prename}file-svg"
+			mv ${plugin}/{,${prename}}${plugin} || exit 1
+			mv {,${prename}}${plugin} || exit 1
+		    fi
+		done
+	)
+}
+
+src_test() {
```
Comment (thesamesam) [inline media-gfx/gimp/gimp-2.99.12.ebuild:220]: You may be able to drop this.

File: media-gfx/gimp/gimp-2.99.12.ebuild [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+	einfo 'Renaming plug-ins to not collide with pre-2.10.6 file layout (bug #664938)...'
+	local prename=gimp-org-
+	(
+		cd "${ED}"/usr/$(get_libdir)/gimp/2.99/plug-ins || exit 1
+		for plugin_slash in $(ls -d1 */); do
+		    plugin=${plugin_slash%/}
+		    if [[ -f ${plugin}/${plugin} ]]; then
+			# NOTE: Folder and file name need to match for Gimp to load that plug-in
+			#       so "file-svg/file-svg" becomes "${prename}file-svg/${prename}file-svg"
+			mv ${plugin}/{,${prename}}${plugin} || exit 1
```
Comment (thesamesam) [inline media-gfx/gimp/gimp-2.99.12.ebuild:213]: Fix indentation

**Review by thesamesam — COMMENTED** [review]
Yeah, please throw in commit(s) to add Python 3.11 to needed packages if their test suites pass.

I'm a bit worried about the SIMD thing you mentioned. We need to make sure it will not try to use additional instructions like e.g. VSX on PPC if the user does not want it.

**Comment by thesamesam** [comment]
cc @ionenwks in case you have any views on meson port, know you've been rocking gimp 3 a bit?

**Comment by ionenwks** [comment]
> cc @ionenwks in case you have any views on meson port, know you've been rocking gimp 3 a bit?

Only gave it a quick look and used a bit, but meson version seems fine to me. And yeah, while I don't use it that much lately, I dropped gimp2 in favor of 3 since Jan 2021. Been happy with how it's been maintained :+1:

**Comment by band-a-prend** [comment]
> And it's seems that MMX and SSE always switched off.

There is was a typo in meson.build: '-mmx', '-sse' instead of '-mmmx', '-msse'. This issue was reported and the fix now is in upstream. The gimp-2.99.12 has this patch in `src_prepare()` and now the support of this options detects by configuration.

But currently there is no possibility to rule such options. I will create request to upstream about it. At least the media-libs/babl uses Meson too and allows to manage such cpu flags.

File: media-gfx/gimp/gimp-2.99.12.ebuild (hunk 6 above)
Comment (band-a-prend) [reply media-gfx/gimp/gimp-2.99.12.ebuild]: Done.

File: media-gfx/gimp/gimp-2.99.12.ebuild (hunk 5 above)
Comment (band-a-prend) [reply media-gfx/gimp/gimp-2.99.12.ebuild]: Removed.

File: media-gfx/gimp/gimp-2.99.12.ebuild (hunk 4 above)
Comment (band-a-prend) [reply media-gfx/gimp/gimp-2.99.12.ebuild]: Dropped (configure.ac parameter).

File: media-gfx/gimp/gimp-2.99.12.ebuild (hunk 3 above)
Comment (band-a-prend) [reply media-gfx/gimp/gimp-2.99.12.ebuild:101]: Oh, I completely missed it with `GNOME2_EAUTORECONF=yes`! 

In update I dropped `gtk-update-icon-cache` (now xdg.eclass is used anyway), `dev-util/intltool` ("NEWS" file mentions: "Full migration from intltool to gettext only."' ), `autoconf`, `automake` and `libtool` (built is ok without last being installed).

Could `test?` and `vala?` conditions stay within `DEPEND`?

File: media-gfx/gimp/gimp-2.99.12.ebuild (hunk 2 above)
Comment (band-a-prend) [reply media-gfx/gimp/gimp-2.99.12.ebuild:15]: Done.

File: media-gfx/gimp/gimp-2.99.12.ebuild (hunk 1 above)
Comment (band-a-prend) [reply media-gfx/gimp/gimp-2.99.12.ebuild]: Done.

**Comment by thesamesam** [comment]
Thanks! I don't see a reason to delay getting this in. It's natural for there to be quirks and bumps along the ride when porting to a new build system.

**Pull request closed without merging**